test/e2e/events_test.go

```diff
@@ -0,0 +1,337 @@
+		// which means we need to wait until 5 Pod retries fail, eventually
+		// triggering a Job failure; the back-off time between the runs is
+		// unfortunately exponential and non-configurable, which makes this test
+		// awfully long. Try to at least make it possible to run in parallel with
+		// other tests.
+		t.Skip("This test currently takes minutes to run, because of exponential backoff in kubernetes")
+	}
+
+	ctx := framework.NewTestCtx(t)
+	defer ctx.Cleanup()
+
+	namespace, err := ctx.GetNamespace()
+	if err != nil {
+		t.Fatalf("could not get namespace: %v", err)
+	}
+	defer DumpJobsLogsOnError(t, framework.Global, namespace)
+	err = framework.AddToFrameworkScheme(apis.AddToScheme, &gitopsv1alpha1.GitOpsConfigList{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	eunomiaURI, found := os.LookupEnv("EUNOMIA_URI")
+	if !found {
+		eunomiaURI = "https://github.com/kohlstechnology/eunomia"
+	}
+	eunomiaRef, found := os.LookupEnv("EUNOMIA_REF")
+	if !found {
+		eunomiaRef = "master"
+	}
+
+	// Step 1: register an event monitor/watcher
+
+	events := make(chan *eventv1beta1.Event, 5)
+	closer, err := watchEvents(events, 5*time.Minute, namespace, "gitops-events-hello-failed")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer closer()
+
+	// Step 2: create a CR with an invalid URI
+
+	gitops := &gitopsv1alpha1.GitOpsConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "GitOpsConfig",
+			APIVersion: "eunomia.kohls.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "gitops-events-hello-failed",
+			Namespace: namespace,
+		},
+		Spec: gitopsv1alpha1.GitOpsConfigSpec{
+			TemplateSource: gitopsv1alpha1.GitConfig{
+				URI:        "https://INVALID!!!",
+				Ref:        eunomiaRef,
+				ContextDir: "URI is already invalid so this value should be irrelevant",
+			},
+			ParameterSource: gitopsv1alpha1.GitConfig{
+				URI:        eunomiaURI,
+				Ref:        eunomiaRef,
+				ContextDir: "test/e2e/testdata/empty-yaml",
+			},
+			Triggers: []gitopsv1alpha1.GitOpsTrigger{
+				{Type: "Change"},
+			},
+			TemplateProcessorImage: "quay.io/kohlstechnology/eunomia-base:dev",
+			ResourceHandlingMode:   "CreateOrMerge",
+			ResourceDeletionMode:   "Delete",
+			ServiceAccountRef:      "eunomia-operator",
+		},
+	}
+	gitops.Annotations = map[string]string{"gitopsconfig.eunomia.kohls.io/initialized": "true"}
+
+	err = framework.Global.Client.Create(context.TODO(), gitops, &framework.CleanupOptions{TestContext: ctx, Timeout: timeout, RetryInterval: retryInterval})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Step 3: Verify events
+
+	// Verify there was an event emitted mentioning that Job failed
+	select {
+	case event := <-events:
+		if event.Reason != "JobFailed" {
+			t.Errorf("got bad event: %v", event)
+		}
+	case <-time.After(3 * time.Minute):
+		t.Errorf("timeout waiting for JobFailed event")
+	}
+}
+
+func watchEvents(events chan<- *eventv1beta1.Event, timeout time.Duration, namespace, gitops string) (closer func(), err error) {
```

Comment:
I think this could be moved into `util.go` in case others want to watch for events in their tests

Reply:
Done, thanks for the idea!

Reply:
Done, thanks for the idea!